docs/COMPLEX_COMPOSITION.md

```diff
@@ -16,7 +16,160 @@ rule <rule name> [when <condition>] {
+1. `Clause 2` is evaluated after complete evaluation of `Clause 1`.
+1. `Clause 2` is evaluated on `LoggingConfiguration` property of all S3 buckets and the result of `Clause 2` is the cumulative result of the evaluation.
+
+### Example 2
+
+```
+let s3_buckets = Resources.*[ Type == "AWS::S3::Bucket" ]
+
+when %s3_buckets !empty {
+  %s3_buckets {
+    Properties.AccessControl exists               << Clause 1 >>
+    Properties.LoggingConfiguration exists        << Clause 2 >>
+  }
+}
+```
+
+The pseudo code in java for the above example is a follows,
+
+```
+boolean isAccessControlExists = true;
+boolean isLoggingConfigurationExists = true;
+
+for each_s3_bucket in s3_buckets {
+    if (each_s3_bucket.Properties.AccessControl == null) {
+        isAccessControlExists = false;
+    }
+
+    if (each_s3_bucket.Properties.LoggingConfiguration == null) {
+        isLoggingConfigurationExists = false;
+    }
+
+    return isAccessControlExists && isLoggingConfigurationExists
+}
+```
+
+In the above example,
+1. `Clause 1` and `Clause 2` are independent statements **within the scope of the query block**, i.e. for each S3 bucket in `%s3_buckets`.
+1. `Clause 1` is evaluated on `AccessControl` property of the first S3 bucket in `%s3_buckets`.
+1. `Clause 2` is evaluated after evaluation of `Clause 1` for that same S3 bucket.
+1. `Clause 2` is evaluated on `LoggingConfiguration` property of that same S3 bucket.
+1. Steps 2 to 4 are executed in that order for the remaining S3 buckets in `%s3_buckets` one after the other until there are no more S3 buckets in `%s3_buckets`.
+
+### Example 1 vs Example 2
+
+There is clearly a difference in the style of execution between Example 1 and Example 2. The clauses in Example 1 can be referred to as **Independent Clause Composition** with respect to S3 buckets in `%s3_buckets` and the clauses in Example 2 can be referred to as **Conjoined Clause Composition** with respect to S3 buckets in `%s3_buckets`. But looking at the pseudo code and the execution order description you can observe that the outcomes of both the examples are the same. However if the conjunction between `Clause 1` and `Clause 2` is switched to disjunction the outcomes would be different. Let us look at Example 1 and Example 2 with conjunction between `Clause 1` and `Clause 2`.
+
+### Example 3 - Example 1 with disjunction between clauses
+
+```
+let s3_buckets = Resources.*[ Type == "AWS::S3::Bucket" ]
+
+when %s3_buckets !empty {
+  %s3_buckets.Properties.AccessControl exists <b>OR</b>            << Clause 1 >>
+  %s3_buckets.Properties.LoggingConfiguration exists        << Clause 2 >>
+}
+```
+
+The pseudo code in java for the above example is a follows,
+
+```
+boolean isAccessControlExists = true;
+boolean isLoggingConfigurationExists = true;
+
+for each_s3_bucket in s3_buckets {
+    if (eachS3Bucket.properties.accessControl == null) {
```

Comment:
Can we keep the casing consistent for P(vs p) properties and A (vs a) accessControl across all code samples?

Reply:
Right! missed it. Will correct it.